### PR TITLE
Parallelize Github unit and integration tests

### LINKED ISSUE #145: Split the CLI integration CI/CD tests out from the matrix run steps to improve test suite speed
## Context

The test suite takes awhile to run, presenting challenges when debugging subtle OS-specific issues.
We want to try what we can to speed it up.

## Task

- [ ] split the CLI integration tests (docker compose steps) into a separate workflow so that they do not have to run with every matrix test.

### LINKED ISSUE #151: macOS Brew recipe for docker-compose is failing to install
## Context

https://github.com/techservicesillinois/awscli-login/actions/runs/5894707628/job/16018161150?pr=149#step:12:62

Our working theory is that the `brew` recipe may no longer be compatible with GitHub actions.
It looks like it is trying to start [Qemu][1] during install, and it is failing. 
We would not expect any kind of VM to work on GitHub actions.

[1]: https://www.qemu.org/

## Tasks

- [ ] update integration tests to only run on Linux and Windows in CI/CD

## Error trace from GitHub action

```bash
Run brew install docker docker-compose
Warning: Treating docker as a formula. For the cask, use homebrew/cask/docker
==> Fetching dependencies for docker: docker-completion
==> Fetching docker-completion
==> Downloading https://ghcr.io/v2/homebrew/core/docker-completion/manifests/24.0.5
==> Downloading https://ghcr.io/v2/homebrew/core/docker-completion/blobs/sha256:b77b6d4980dd2c9345cf64b9a790c0ac177358195d19202a06112b19b1533a1d
==> Fetching docker
==> Downloading https://ghcr.io/v2/homebrew/core/docker/manifests/24.0.5
==> Downloading https://ghcr.io/v2/homebrew/core/docker/blobs/sha256:25a7049a99a9a2e8a9729a2884a177e5fb692f160efd07db5be26a44d5174c66
==> Fetching docker-compose
==> Downloading https://ghcr.io/v2/homebrew/core/docker-compose/manifests/2.20.3
==> Downloading https://ghcr.io/v2/homebrew/core/docker-compose/blobs/sha256:174117b0c540c90033c0835e493f19bf96d5bc97a28dd12f0b2a19e179aa8638
==> Installing dependencies for docker: docker-completion
==> Installing docker dependency: docker-completion
==> Pouring docker-completion--24.0.5.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/docker-completion/24.0.5: 9 files, 337.8KB
==> Installing docker
==> Pouring docker--24.0.5.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/docker/24.0.5: 13 files, 32.2MB
==> Pouring docker-compose--2.20.3.monterey.bottle.tar.gz
==> Caveats
Compose is now a Docker plugin. For Docker to find this plugin, symlink it:
  mkdir -p ~/.docker/cli-plugins
  ln -sfn /usr/local/opt/docker-compose/bin/docker-compose ~/.docker/cli-plugins/docker-compose
==> Summary
🍺  /usr/local/Cellar/docker-compose/2.20.3: 6 files, 57.5MB
==> Caveats
==> docker-compose
Compose is now a Docker plugin. For Docker to find this plugin, symlink it:
  mkdir -p ~/.docker/cli-plugins
time="2023-08-18T16:37:03Z" level=info msg="starting colima"
  ln -sfn /usr/local/opt/docker-compose/bin/docker-compose ~/.docker/cli-plugins/docker-compose
time="2023-08-18T16:37:03Z" level=info msg="runtime: docker"
time="2023-08-18T16:37:03Z" level=info msg="preparing network ..." context=vm
time="2023-08-18T16:37:03Z" level=info msg="creating and starting ..." context=vm
time="2023-08-18T16:37:03Z" level=info msg="Terminal is not available, proceeding without opening an editor"
time="2023-08-18T16:37:03Z" level=info msg="Attempting to download the image" arch=x86_64 digest="sha512:f761b807fe9ba345968df72c07f8c5abcae0c4a44976fe5595c0ff748ef693841221a70e663986c700b027cea32b7cac24d5490d4c721593c39f2b8840c362a2" location="https://github.com/abiosoft/alpine-lima/releases/download/colima-v0.5.5/alpine-lima-clm-3.18.0-x86_64.iso"
Downloading the image (alpine-lima-clm-3.18.0-x86_64.iso)

315.00 MiB / 315.00 MiB (100.00%) ? p/stime="2023-08-18T16:37:05Z" level=info msg="Downloaded the image from \"[https://github.com/abiosoft/alpine-lima/releases/download/colima-v0.5.5/alpine-lima-clm-3.18.0-x86_64.iso\](https://github.com/abiosoft/alpine-lima/releases/download/colima-v0.5.5/alpine-lima-clm-3.18.0-x86_64.iso/)""
time="2023-08-18T16:37:09Z" level=info msg="[hostagent] Starting QEMU (hint: to watch the boot progress, see \"/Users/runner/.lima/colima/serial*.log\")"
time="2023-08-18T16:37:09Z" level=info msg="SSH Local Port: 49228"
time="2023-08-18T16:37:09Z" level=info msg="[hostagent] Waiting for the essential requirement 1 of 5: \"ssh\""
time="2023-08-18T16:37:09Z" level=info msg="[hostagent] Driver stopped due to error: \"exit status 255\""
time="2023-08-18T16:37:09Z" level=info msg="[hostagent] Shutting down the host agent"
time="2023-08-18T16:37:09Z" level=warning msg="[hostagent] failed to exit SSH master"
time="2023-08-18T16:37:09Z" level=info msg="[hostagent] Shutting down QEMU with ACPI"
time="2023-08-18T16:37:09Z" level=warning msg="[hostagent] failed to open the QMP socket \"/Users/runner/.lima/colima/qmp.sock\", forcibly killing QEMU"
time="2023-08-18T16:37:09Z" level=info msg="[hostagent] QEMU has already exited"
time="2023-08-18T16:37:09Z" level=fatal msg="exiting, status={Running:false Degraded:false Exiting:true Errors:[] SSHLocalPort:0} (hint: see \"/Users/runner/.lima/colima/ha.stderr.log\")"
time="2023-08-18T16:37:09Z" level=fatal msg="error starting vm: error at 'creating and starting': exit status 1"
Error: Process completed with exit code 1.
```

### PR DESCRIPTION
**Pull request by ddriddle**
Closes #145
Closes #151

### CODE REVIEW

File: .github/workflows/cicd.yml [hunk 1]
```diff
@@ -154,8 +146,63 @@ jobs:
...
+      run: echo "::set-output name=dir::$(python -m pip cache dir)"
+
+    - name: Save pip cache
+      uses: actions/cache@v3
+      with:
+        path: ${{ steps.pip-cache.outputs.dir }}
+        key: ${{ runner.os }}-pip-${{ secrets.CACHE_VERSION }}
+
+    - name: Install system dependencies
+      if: matrix.os == 'ubuntu-latest' && steps.cache.outputs.valid != 'true'
```
Comment (edthedev) [inline .github/workflows/cicd.yml:190]:
```suggestion
      if: steps.cache.outputs.valid != 'true'
```
Per discussion

File: .github/workflows/cicd.yml [hunk 2]
```diff
@@ -154,8 +146,63 @@ jobs:
...
+    - name: Install system dependencies
+      if: steps.cache.outputs.valid != 'true'
+      run: |
+        sudo apt-get install -y libxml2-dev libxslt-dev
+
+    - name: Install system dependencies
+      if: matrix.os == 'macos-latest'
+      run: |
+        brew install docker docker-compose
+        colima start
```
Comment (edthedev) [inline .github/workflows/cicd.yml:198]:
```suggestion
```
Remove mac bit per discussion

File: .github/workflows/cicd.yml [hunk 3]
```diff
@@ -154,8 +146,57 @@ jobs:
...
-    - name: Install (local)
-      run: make .install
+    - name: Archive coverage report
+      uses: actions/upload-artifact@v3
+      with:
+        name: coverage-${{ runner.os }}-py${{ matrix.python-version }}
+        path: htmlcov
+
+  integration_tests:
+    runs-on: ${{ matrix.os }}
```
Comment (edthedev) [inline .github/workflows/cicd.yml:156]:
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/cicd.yml [hunk 4]
```diff
@@ -154,8 +146,57 @@ jobs:
...
+        name: coverage-${{ runner.os }}-py${{ matrix.python-version }}
+        path: htmlcov
+
+  integration_tests:
+    runs-on: ubuntu-latest
+    needs: build
+
+    strategy:
+      matrix:
+        os: ubuntu-latest
```
Comment (edthedev) [inline .github/workflows/cicd.yml:161]:
```suggestion
```
No matrix here.

**Review by mpitcel — APPROVED** [review]

**Review by tzturner — APPROVED** [review]

**Pull request merged**